### DISCUSSION TITLE
[Network] Allow plugins to simulate various network conditions

### DISCUSSION THREAD

**Discussion by abhinavsingh** [abhinavsingh/proxy.py#794] [Ideas]
**Describe the solution you'd like**
Add options to simulate various network types (per-client and / or globally) e.g. `2g`, `3g`, .... Custom profiles can be added by specifying packet drop, latency and other metrics via plugins.

Starter profiles can be found in [webpagetest](https://github.com/WPO-Foundation/webpagetest/blob/master/www/settings/connectivity.ini.sample) project repo.

**Comment by abhinavsingh** [comment]
Also see https://github.com/h2non/toxy#poisons for some interesting examples and use-cases.

**Comment by sakurai-youhei** [comment]
@abhinavsingh Let me advertise :) I wrote [one plugin](https://github.com/sakurai-youhei/proxy.py-uxspoilers-plugin) which would simulate a poor network to some extent. My main purpose is not the network simulation but spoiling a user experience, though. Hope it would help those who come here for the same reason in some future.

**Reply by abhinavsingh** [reply]
> @abhinavsingh Let me advertise :) I wrote [one plugin](https://github.com/sakurai-youhei/proxy.py-uxspoilers-plugin) which would simulate a poor network to some extent. My main purpose is not the network simulation but spoiling a user experience, though. Hope it would help those who come here for the same reason in some future.

Can you bring it in as a plugin?  Send in a PR.  We can formalize it to match various network profiles (jitter/loss/drop).  Add flags to let users enable network shapers or let plugins control the network shapers per request.
